hookutils: collect_submodules: do not recurse into filtered subpackages

rokm opened the pull request:
Reorganize the `collect_submodules` code so that each (sub)package scan step is performed separately, and the filter function is applied to the results before recursing into the child subpackages.

This ensures that we do not attempt to recurse into subpackages that are filtered out (i.e., packages that are known to be un-importable and generate a warning/error depending on the on_error setting).

Closes #6846.

rokm committed:
hookutils: collect_submodules: do not recurse into filtered subpackages

Reorganize the collect_submodules code so that each (sub)package
scan step is performed separately, and the filter function is
applied to the results before recursing into the child subpackages.

This ensures that we do not attempt to recurse into subpackages
that are filtered out (i.e., packages that are known to be
un-importable and generate a warning/error depending on the
on_error setting).

rokm commented:
There are two more aspects that we should probably address at this point. I did not change them yet because we have tests that explicitly test for the current behavior, so they would also need to be adjusted.

1. Should passing an importable module to the function return list with that module's name, or an empty list as it does now? `collect_submodules('mypkg')` includes `mypkg` in the returned list, so it might be reasonable for `collect_submodules('os')` to return `['os']` as well. This came up in #6824.

2. If a subpackage cannot be imported, should it be added to the list of submodules (which is the current behavior) or not? #6846 argues that it would make more sense to exclude such subpackages (note that this applies only to sub-packages, as we do not try to directly import individual submodules). I don't have a strong opinion on this, but if we are going to change the behavior, we should do it now.

bwoodsend commented:
My answers to both 1 and 2 are that the current behaviours are quirky and the proposed changes make more sense.

rokm committed:
hookutils: collect_submodules: if passed a module, return its name

If passed a name of importable module instead of a package,
include its name in the returned list instead of returning an
empty list. This makes behavior consistent with behavior in the
case of a package without submodules.

rokm committed:
hookutils: collect_submodules: exclude un-importable subpackages

If a subpackage cannot be imported, do not include it in the
returned list of modules. We cannot recurse into such a subpackage
and therefore its submodules cannot be determined, so it makes
little sense to include the subpackage itself.

rokm commented:
I suppose the counter argument for 2. is that for un-importable subpackage, the program might be expecting some other exception type rather than the `ModuleNotFound` error. Although as we fail to collect submodules of such a subpackage, its init might end up raising a `ModuleNotFound` before raising the other error...

bwoodsend commented:
You mean something like:
```python
# Check if `optional_package` is available. May raise an XError() if this
# platform/hardware/star alignment is not to its liking.
try:
    import optional_package
except XError:
    optional_package = None
```
In which case, omitting `optional_package` because it fails to import will lead to a `ModuleNotFoundError` instead of an `XError` when frozen and therefore not be caught by the `except XError` branch?

rokm commented:
Something like that, yeah. But with additional complexity of programmatic importing of sub-packages (plugins, for example?), which warrants the use of `collect_submodules` on the package.

bwoodsend commented:
Unless the unimportable (sub)module has no child submodules, such a case would be pretty much doomed anyway though since PyInstaller will not have collected the child submodules, right?

rokm commented:
> Unless the unimportable (sub)module has no child submodules, such a case would be pretty much doomed anyway though since PyInstaller will not have collected the child submodules, right?

Yeah, the unimportable subpackage might raise `ModuleNotFoundError` due to its modules not being collected anyway even if we collect the subpackage itself. So in the grand scheme of things, it probably does not make much of a difference.

rokm commented:
I've squashed the first three commits (original change, test modification, and regression fix-up) together to make this cleaner to merge...
